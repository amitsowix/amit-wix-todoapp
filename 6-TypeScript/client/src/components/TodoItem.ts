import { styles } from "../helpers/jss";

export class TodoItem {

    id: string;
    text: string;
    isChecked: Boolean;
    todoItemContainer: HTMLElement;
    checkbox: HTMLInputElement;
    textElement: HTMLElement;
    actionsContainer: HTMLElement;
    editIcon: HTMLImageElement;
    deleteIcon: HTMLImageElement;
    isInEditMode: Boolean;
    editInput: HTMLInputElement;
    submitButton: HTMLElement;

    constructor(id: string, text: string, isChecked: Boolean) {
        this.id = id;
        this.text = text;
        this.isChecked = isChecked;
        this.isInEditMode = false;
        this.createElements();
    }

    createTodoItemContainer: Function = () => {
        this.todoItemContainer = document.createElement('div');
        this.todoItemContainer.className = styles.classes.todoItem;
    }


    onToggleCheckbox: Function = () => {
        this.textElement.className = this.checkbox.checked ? styles.classes.todoItemTextChecked : styles.classes.todoItemText;
        //UPDATELIST
    }

    createCheckbox: Function = () => {
        this.checkbox = document.createElement('input');
        this.checkbox.type = 'checkbox';
        this.checkbox.checked = this.isChecked ? true : false;
        this.checkbox.className = styles.classes.checkbox;
        this.checkbox.addEventListener('click', () => this.onToggleCheckbox());
    }

    createTextElement: Function = () => {
        this.textElement = document.createElement('div');
        this.textElement.innerText = this.text;
        this.textElement.className = this.checkbox.checked ? styles.classes.todoItemTextChecked : styles.classes.todoItemText;
    }

    createEditIcon: Function = () => {
        this.editIcon = document.createElement('img');
        this.editIcon.className = styles.classes.icon;
        this.editIcon.src = "./edit.png";
    }

    createEditInput: Function = () => {
        this.editInput = document.createElement('input');
        this.editInput.className = styles.classes.input;
        this.submitButton = document.createElement('div');
        this.submitButton.innerText = "GO";
        this.submitButton.className = styles.classes.submitButton;
    }

    createDeleteIcon: Function = () => {
        this.deleteIcon = document.createElement('img');
        this.deleteIcon.className = styles.classes.icon;
        this.deleteIcon.src = "./delete.png";
    }

    createActionsContainer: Function = () => {
        this.actionsContainer = document.createElement('div');
        this.actionsContainer.className = styles.classes.actionsContainer;
        this.createDeleteIcon();
        this.createEditIcon();
        this.createEditInput();
        this.actionsContainer.appendChild(this.editIcon);
        this.actionsContainer.appendChild(this.editInput);
        this.actionsContainer.appendChild(this.submitButton);
        this.actionsContainer.appendChild(this.deleteIcon);
    }

    showEditInput: Function = () => {
        this.editInput.style.display = 'block';
        this.submitButton.style.display = 'block';
        this.isInEditMode = true;
        this.editInput.focus();
    }
    
    hideEditInput: Function = () => {
        this.editInput.style.display = 'none';
        this.submitButton.style.display = 'none';
        this.isInEditMode = false;
    }   

    onEditInputSubmit: Function = () => {
        this.textElement.innerText = this.editInput.value;
        this.editInput.value = "";
        this.hideEditInput();
        //UPDATELIST
    }

    setEditEvents: Function = () => {
        this.editIcon.addEventListener('click', () => {
            this.isInEditMode ? this.hideEditInput() : this.showEditInput();
        })
        this.editInput.addEventListener('keyup', (event) => {
            if (event.key === "Enter" && this.editInput.value !== ""){
                this.onEditInputSubmit();
            }
        })
        this.submitButton.addEventListener('click', () => {
            if (this.editInput.value !== ""){
                this.onEditInputSubmit();
            }
        })
    }

    setDeleteEvents: Function = () => {
        this.deleteIcon.addEventListener('click', () => {
            //DELETEITEM
        })
    }

    createElements: Function = () => {
        this.createTodoItemContainer();
        this.createCheckbox();
        this.createTextElement();
        this.createActionsContainer();
        this.setEditEvents();
        this.setDeleteEvents();
        this.todoItemContainer.appendChild(this.checkbox);
        this.todoItemContainer.appendChild(this.textElement);
        this.todoItemContainer.appendChild(this.actionsContainer);
    }
}