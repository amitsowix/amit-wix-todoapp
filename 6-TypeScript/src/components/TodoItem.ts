import { styles } from "../helpers/jss";
import serverApi from "../helpers/serverapi";

export class TodoItem {

    id: string;
    text: string;
    isChecked: Boolean;
    todoItemContainer: HTMLElement;
    checkbox: HTMLInputElement;
    textElement: HTMLElement;
    actionsContainer: HTMLElement;
    editIcon: HTMLImageElement;
    deleteIcon: HTMLImageElement;
    isInEditMode: Boolean;
    editInput: HTMLInputElement;
    submitButton: HTMLElement;
    deleteItemFromList: (todoItem: HTMLElement) => void;

    constructor(id: string, text: string, isChecked: Boolean, deleteItemFromList: (todoItem: HTMLElement) => void) {
        this.id = id;
        this.text = text;
        this.isChecked = isChecked;
        this.isInEditMode = false;
        this.deleteItemFromList = deleteItemFromList;
        this.createElements();
    }

    createTodoItemContainer = (): void => {
        this.todoItemContainer = document.createElement('div');
        this.todoItemContainer.className = styles.classes.todoItem;
    }

    updateTodoItem = (): void => {
        serverApi.editTodoItem(this.id, {id: this.id, text: this.text, isChecked: this.isChecked});
    }


    onToggleCheckbox = (): void => {
        this.textElement.className = this.checkbox.checked ? styles.classes.todoItemTextChecked : styles.classes.todoItemText;
        this.isChecked = this.checkbox.checked;
        this.updateTodoItem();
    }

    createCheckbox = (): void => {
        this.checkbox = document.createElement('input');
        this.checkbox.type = 'checkbox';
        this.checkbox.checked = this.isChecked ? true : false;
        this.checkbox.className = styles.classes.checkbox;
        this.checkbox.addEventListener('click', () => this.onToggleCheckbox());
    }

    createTextElement = (): void => {
        this.textElement = document.createElement('div');
        this.textElement.innerText = this.text;
        this.textElement.className = this.checkbox.checked ? styles.classes.todoItemTextChecked : styles.classes.todoItemText;
    }

    createEditIcon = (): void => {
        this.editIcon = document.createElement('img');
        this.editIcon.className = styles.classes.icon;
        this.editIcon.src = "./edit.png";
    }

    createEditInput = (): void => {
        this.editInput = document.createElement('input');
        this.editInput.className = styles.classes.input;
        this.submitButton = document.createElement('div');
        this.submitButton.innerText = "GO";
        this.submitButton.className = styles.classes.submitButton;
    }

    createDeleteIcon = (): void => {
        this.deleteIcon = document.createElement('img');
        this.deleteIcon.className = styles.classes.icon;
        this.deleteIcon.src = "./delete.png";
    }

    createActionsContainer = (): void => {
        this.actionsContainer = document.createElement('div');
        this.actionsContainer.className = styles.classes.actionsContainer;
        this.createDeleteIcon();
        this.createEditIcon();
        this.createEditInput();
        this.actionsContainer.appendChild(this.editIcon);
        this.actionsContainer.appendChild(this.editInput);
        this.actionsContainer.appendChild(this.submitButton);
        this.actionsContainer.appendChild(this.deleteIcon);
    }

    showEditInput = (): void => {
        this.editInput.style.display = 'block';
        this.submitButton.style.display = 'block';
        this.isInEditMode = true;
        this.editInput.focus();
    }
    
    hideEditInput = (): void => {
        this.editInput.style.display = 'none';
        this.submitButton.style.display = 'none';
        this.isInEditMode = false;
    }   

    onEditInputSubmit = (): void => {
        this.textElement.innerText = this.editInput.value;
        this.text = this.editInput.value;
        this.editInput.value = "";
        this.hideEditInput();
        this.updateTodoItem();
    }

    setEditEvents = (): void => {
        this.editIcon.addEventListener('click', () => {
            this.isInEditMode ? this.hideEditInput() : this.showEditInput();
        })
        this.editInput.addEventListener('keyup', (event) => {
            if (event.key === "Enter" && this.editInput.value !== ""){
                this.onEditInputSubmit();
            }
        })
        this.submitButton.addEventListener('click', () => {
            if (this.editInput.value !== ""){
                this.onEditInputSubmit();
            }
        })
    }

    setDeleteEvents = (): void => {
        this.deleteIcon.addEventListener('click', () => {
            serverApi.deleteTodoItem(this.id);
            this.deleteItemFromList(this.todoItemContainer);
        })
    }

    createElements = (): void => {
        this.createTodoItemContainer();
        this.createCheckbox();
        this.createTextElement();
        this.createActionsContainer();
        this.setEditEvents();
        this.setDeleteEvents();
        this.todoItemContainer.appendChild(this.checkbox);
        this.todoItemContainer.appendChild(this.textElement);
        this.todoItemContainer.appendChild(this.actionsContainer);
    }
}