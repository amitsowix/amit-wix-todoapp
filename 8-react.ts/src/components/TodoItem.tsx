import {FC, useState} from 'react';
import {createUseStyles} from 'react-jss'
import {Classes} from 'jss';
import {TodoItemDTO, UUID} from "../../../common/interfaces";
import editIcon from "../assets/edit.png";
import deleteIcon from "../assets/delete.png";
import Input from "./Input";

interface TodoItemProps {
    id: UUID;
    text: string;
    isChecked: boolean;
    editTodoItem: (item: TodoItemDTO) => void;
    deleteTodoItem: (id: string) => void;
}

const TodoItem: FC<TodoItemProps> = (props) : JSX.Element => {
    const classes: Classes = useStyles();

    const [text, setText] = useState(props.text);
    const [isChecked, setIsChecked] = useState(props.isChecked);
    const [isInEditMode, setEditMode] = useState(false);

    return (
        <div className={classes.todoItem}>
            <input 
                type={'checkbox'} 
                className={classes.checkbox}
                onChange={(event)=>{
                    setIsChecked(event.target.checked);
                    props.editTodoItem({id: props.id, text, isChecked: event.target.checked});
                }} 
                checked={isChecked}/>
            <div className={isChecked ? classes.todoItemTextChecked : classes.todoItemText}>
                {text}
            </div>
            <img src={editIcon} className={classes.icon} onClick={()=>setEditMode(!isInEditMode)}/>
            {isInEditMode ? <Input onInputSubmit={(newValue)=>{
                setText(newValue);
                setEditMode(false);
                props.editTodoItem({id: props.id, text: newValue, isChecked});
            }}/> : null}
            <img src={deleteIcon} className={classes.icon} onClick={()=>{props.deleteTodoItem(props.id)}}/>
        </div>
    )            
}

const useStyles: () => Classes = createUseStyles({
    icon: {
        height: "30px",
        marginLeft: "10px",
        cursor: "pointer"
    },
    
    todoItem: {
        display: "flex",
        alignItems: "flex-end",
        marginTop: "30px"
    },
    
    todoItemText: {
        marginLeft: "10px",
        fontSize: "20px" 
    },
    todoItemTextChecked :{
        marginLeft: "10px",
        fontSize: "20px",
        textDecoration: "line-through"
    },
    checkbox: { 
        height: "20px", 
        width: "20px" 
    },
    listContainer: { 
        overflow: "scroll", 
        height: "44%" 
    },
    actionsContainer: { 
        display: "flex" 
    }
});

export default TodoItem;