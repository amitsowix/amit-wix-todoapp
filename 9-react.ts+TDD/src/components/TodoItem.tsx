import React, {FC, useState} from 'react';
import {createUseStyles} from 'react-jss'
import {Classes} from 'jss';
import {TodoItemDTO, UUID} from "../../../common/interfaces";
import editIcon from "../assets/edit.png";
import deleteIcon from "../assets/delete.png";
import Input from "./Input";

interface TodoItemProps {
    id: UUID;
    text: string;
    isChecked: boolean;
    editTodoItem: (item: TodoItemDTO) => void;
    deleteTodoItem: (id: string) => void;
}

const TodoItem: FC<TodoItemProps> = (props) : JSX.Element => {
    const classes: Classes = useStyles();

    const [text, setText] = useState(props.text);
    const [isChecked, setIsChecked] = useState(props.isChecked);
    const [isInEditMode, setEditMode] = useState(false);

    return (
        <div data-hook={'todo-item'} className={classes.todoItem}>
            <input 
                name = {'checkbox-' + props.id}
                type={'checkbox'} 
                className={classes.checkbox}
                onChange={(event)=>{
                    setIsChecked(event.target.checked);
                    props.editTodoItem({id: props.id, text, isChecked: event.target.checked});
                }} 
                checked={isChecked}/>
            <div data-hook={'todo-text'} className={isChecked ? classes.todoItemTextChecked : classes.todoItemText}>
                {text}
            </div>
            <img data-hook={'edit-icon'} id={'edit-icon-' + props.id} src={editIcon} className={classes.icon} onClick={()=>setEditMode(!isInEditMode)}/>
            {isInEditMode ? <Input name={'edit-input'} onInputSubmit={(newValue)=>{
                setText(newValue);
                setEditMode(false);
                props.editTodoItem({id: props.id, text: newValue, isChecked});
            }}/> : null}
            <img data-hook={'delete-icon'} id={'delete-icon-' + props.id} src={deleteIcon} className={classes.icon} onClick={()=>{props.deleteTodoItem(props.id)}}/>
        </div>
    )            
}

const useStyles: () => Classes = createUseStyles({
    icon: {
        height: "30px",
        marginLeft: "10px",
        cursor: "pointer"
    },
    
    todoItem: {
        display: "flex",
        alignItems: "flex-end",
        marginTop: "30px"
    },
    
    todoItemText: {
        marginLeft: "10px",
        fontSize: "20px" 
    },
    todoItemTextChecked :{
        marginLeft: "10px",
        fontSize: "20px",
        textDecoration: "line-through"
    },
    checkbox: { 
        height: "20px", 
        width: "20px" 
    },
    listContainer: { 
        overflow: "scroll", 
        height: "44%" 
    },
    actionsContainer: { 
        display: "flex" 
    }
});

export default TodoItem;