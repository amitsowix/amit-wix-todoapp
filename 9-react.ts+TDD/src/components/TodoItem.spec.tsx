import React from 'react';
import renderer, {ReactTestRenderer, ReactTestRendererJSON} from 'react-test-renderer';
import {screen, render} from '@testing-library/react';
import TodoItem from './TodoItem';
import {mount, ReactWrapper} from 'enzyme';

describe('Test Render', () => {
    it('should render properly', () => {
        const component = driver.when.mount();

        const tree: ReactTestRendererJSON = component.toJSON() as ReactTestRendererJSON;
        expect(tree).toMatchSnapshot();
    })

    it('should be of type div', () => {
        const component: ReactTestRenderer = driver.when.mount();

        const tree: ReactTestRendererJSON = component.toJSON() as ReactTestRendererJSON;
        expect(tree.type).toMatch('div');
    })

    it('should render todoitem text', () => {
        driver.when.render();

        expect(driver.get.todoText()).toBeInTheDocument();
    })
    
    it('should mark the checkbox as unchecked', () => {
        const wrapper: ReactWrapper = driver.when.enzymeMount(undefined, undefined, false);

        const checkbox = driver.get.checkbox(wrapper);
        expect(checkbox.props().checked).toEqual(false);
    })

    it('should mark the checkbox as checeked', () => {
        const wrapper: ReactWrapper = driver.when.enzymeMount(undefined, undefined, true);

        const checkbox = driver.get.checkbox(wrapper);
        expect(checkbox.props().checked).toEqual(true);
    })

    it('input shoudn\'t show', () => {
        const wrapper: ReactWrapper = driver.when.enzymeMount();

        expect(driver.get.input(wrapper).exists()).toBeFalsy();
        expect(driver.get.submitButton(wrapper).exists()).toBeFalsy();
    })
})


describe('Input show/hide', () => {
    it('should show/hide input on click', () => {
        const wrapper: ReactWrapper = driver.when.enzymeMount();
        const editIcon: ReactWrapper = driver.get.editIcon(wrapper);
        driver.when.click(editIcon);

        const input = driver.get.input(wrapper);
        const submitButton = driver.get.submitButton(wrapper);
        expect(input.exists()).toBeTruthy();
        expect(submitButton.exists()).toBeTruthy();

        driver.when.click(editIcon);

        expect(input).not.toBeInTheDocument;
        expect(submitButton).not.toBeInTheDocument;
    })
})



describe('Todo item actions', () => {
    it('should delete the item', () => {
        const mockCallBack = jest.fn();
        const wrapper: ReactWrapper = driver.when.enzymeMount(undefined, mockCallBack);
        const deleteIcon = driver.get.deleteIcon(wrapper);
        driver.when.click(deleteIcon);

        expect(mockCallBack.mock.calls.length).toEqual(1);
    });
})


const driver = {
    when: {
        mount: (editCallback: () => void = () => {}, deleteCallback: () => void = () => {}, isChecked: boolean = false): ReactTestRenderer => {
            const component: ReactTestRenderer = renderer.create(
                <TodoItem id={'mock'} text={'mock'} isChecked={isChecked} editTodoItem={editCallback} deleteTodoItem={deleteCallback}/>
            )            
            return component;
        },
        render: (editCallback: () => void = () => {}, deleteCallback: () => void = () => {}, isChecked: boolean = false): void => {
            render(
                <TodoItem id={'mock'} text={'mock'} isChecked={isChecked} editTodoItem={editCallback} deleteTodoItem={deleteCallback}/>             
            );
        },
        enzymeMount: (editCallback: () => void = () => {}, deleteCallback: () => void = () => {}, isChecked: boolean = false): ReactWrapper => {
            const wrapper = mount(<TodoItem id={'mock'} text={'mock'} isChecked={isChecked} editTodoItem={editCallback} deleteTodoItem={deleteCallback}/>);
            return wrapper;
        },
        click: (element: ReactWrapper): void => {
            element.simulate('click');
        },
    },
    get: {
      todoText: () => screen.getByText('mock'),
      submitButton: (wrapper: ReactWrapper) => wrapper.find('div[data-hook="submit-button"]'),
      input: (wrapper: ReactWrapper) => wrapper.find('input[name="edit-input"]'),
      checkbox: (wrapper: ReactWrapper) => wrapper.find('input').at(0),
      editIcon: (wrapper: ReactWrapper) => wrapper.find('img').at(0),
      deleteIcon: (wrapper: ReactWrapper) => wrapper.find('img').at(1)
    }
}