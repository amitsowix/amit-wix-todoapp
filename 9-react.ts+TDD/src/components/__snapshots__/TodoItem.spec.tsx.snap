// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Render should render properly 1`] = `
<div
  className="todoItem-0-2-2"
  data-hook="todo-item"
>
  <input
    checked={false}
    className="checkbox-0-2-5"
    name="checkbox-mock"
    onChange={[Function]}
    type="checkbox"
  />
  <div
    className="todoItemText-0-2-3"
    data-hook="todo-text"
  >
    mock
  </div>
  <img
    className="icon-0-2-1"
    data-hook="edit-icon"
    id="edit-icon-mock"
    onClick={[Function]}
    src="test-file-stub"
  />
  <img
    className="icon-0-2-1"
    data-hook="delete-icon"
    id="delete-icon-mock"
    onClick={[Function]}
    src="test-file-stub"
  />
</div>
`;
